core/model/src/main/java/org/eclipse/rdf4j/model/util/RDFContainers.java

```diff
@@ -0,0 +1,160 @@
+			throws ModelException {
+		Objects.requireNonNull(consumer, "consumer may not be null");
+		Objects.requireNonNull(m, "input model may not be null");
+
+		ValueFactory vf = SimpleValueFactory.getInstance();
+
+		GetStatementOptional statementSupplier = (s, p, o, c) -> m.filter(s, p, o, c).stream().findAny();
+		Function<String, Supplier<ModelException>> exceptionSupplier = Models::modelException;
+
+		// TODO add proper documentation
+		Pattern annotatedMembershipPredicatePattern = Pattern.compile("^" + vf.createIRI(RDF.NAMESPACE, "_") + "[1-9][0-9]*$");
+
+		extract(containerType, statementSupplier, container, st -> {
+			if (RDFS.MEMBER.equals(st.getPredicate()) ||
+				annotatedMembershipPredicatePattern.matcher(st.getPredicate().toString()).matches()) {
+				consumer.accept(st.getObject());
+			}
+		}, exceptionSupplier, contexts);
+	}
+
+	public static <C extends Collection<Statement>> C getContainer(IRI containerType, Model sourceModel, Resource head, C sink,
+			Resource... contexts) {
+		Objects.requireNonNull(sourceModel, "input model may not be null");
+		extract(containerType, sourceModel, head, st -> sink.add(st), contexts);
+		return sink;
+	}
+
+	public static void extract(IRI containerType, Model sourceModel, Resource container, Consumer<Statement> consumer, Resource... contexts) {
+		Objects.requireNonNull(sourceModel, "source model may not be null");
+		GetStatementOptional statementSupplier = (s, p, o,
+				c) -> ((Model) sourceModel).filter(s, p, o, c).stream().findAny();
+		extract(containerType, statementSupplier, container, consumer, Models::modelException, contexts);
+	}
+
+	public static <E extends RDF4JException> void extract(IRI containerType, GetStatementOptional statementSupplier, Resource container,
+														  Consumer<Statement> collectionConsumer, Function<String, Supplier<E>> exceptionSupplier,
+														  Resource... contexts) throws E {
+		OpenRDFUtil.verifyContextNotNull(contexts);
+		Objects.requireNonNull(container, "list head may not be null");
+		Objects.requireNonNull(collectionConsumer, "collection consumer may not be null");
+
+		ValueFactory vf = SimpleValueFactory.getInstance();
+
+		Resource current = container;
+		final Set<Value> encountered = new HashSet<>();
+
+		for (int annotatedMembershipPropertyCounter = 1; true ; annotatedMembershipPropertyCounter++) {
+
+
+			IRI annotatedMembershipPredicate = getAnnotatedMemberPredicate(vf, annotatedMembershipPropertyCounter);
+			if (statementSupplier.get(container, annotatedMembershipPredicate, null, contexts).equals(Optional.empty())) {
+				break;
+			}
+			Statement statement = statementSupplier.get(container, annotatedMembershipPredicate, null, contexts).get();
+
+			if (containerType.equals(RDF.ALT)) {
+				if (encountered.contains(statement.getObject())) {
+					throw exceptionSupplier.apply("rdf:alt cannot contain duplicate values").get();
+				}
+				encountered.add(statement.getObject());
+			}
```

Comment:
I'd actually leave this check out, completely. It may seem logical to test this, but there is nothing in the RDF specifications that says anything about the members of an `rdf:Alt` container having to be distinct.